SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp

```diff
@@ -687,18 +711,36 @@ FRPCErrorInfo USpatialSender::SendRPC(const FPendingRPCParams& Params)
 	}
 }
 
-bool USpatialSender::SendRingBufferedRPC(UObject* TargetObject, UFunction* Function, const SpatialGDK::RPCPayload& Payload,
-										 USpatialActorChannel* Channel, const FUnrealObjectRef& TargetObjectRef)
+bool USpatialSender::SendCrossServerRPC(UObject* TargetObject, const SpatialGDK::RPCSender& Sender, UFunction* Function,
+										const SpatialGDK::RPCPayload& Payload, USpatialActorChannel* Channel,
+										const FUnrealObjectRef& TargetObjectRef)
 {
 	const FRPCInfo& RPCInfo = ClassInfoManager->GetRPCInfo(TargetObject, Function);
-	const EPushRPCResult Result =
-		RPCService->PushRPC(TargetObjectRef.Entity, RPCInfo.Type, Payload, Channel->bCreatedEntity, TargetObject, Function);
+	const USpatialGDKSettings* Settings = GetDefault<USpatialGDKSettings>();
+	const bool bHasValidSender = Sender.Entity != SpatialConstants::INVALID_ENTITY_ID;
 
-	if (Result == EPushRPCResult::Success)
+	check(Settings->CrossServerRPCImplementation == ECrossServerRPCImplementation::RoutingWorker);
+	if (bHasValidSender)
 	{
-		FlushRPCService();
+		return SendRingBufferedRPC(TargetObject, Sender, Function, Payload, Channel, TargetObjectRef);
 	}
 
+	return false;
+}
+
+bool USpatialSender::SendRingBufferedRPC(UObject* TargetObject, const SpatialGDK::RPCSender& Sender, UFunction* Function,
+										 const SpatialGDK::RPCPayload& Payload, USpatialActorChannel* Channel,
+										 const FUnrealObjectRef& TargetObjectRef)
+{
+	const FRPCInfo& RPCInfo = ClassInfoManager->GetRPCInfo(TargetObject, Function);
+	const EPushRPCResult Result =
+		RPCService->PushRPC(TargetObjectRef.Entity, Sender, RPCInfo.Type, Payload, Channel->bCreatedEntity, TargetObject, Function);
+
+	// if (Result == EPushRPCResult::Success)
+	//{
+	//	FlushRPCService();
+	//}
```

Comment:
```suggestion
```

Reply:
That's right, it should be removed, but that could be considered as a controversial change (more latency), so I'm leaving it there for discussion.
But the crux of it is that if you ever sent a RPC alongside modifying your properties, making the remote side receive the RPC before you had the opportunity of replicating your properties as well might cause issues and unexpected/broken invariants. So that change is likely the correct thing to do.

Reply:
Is this not also the case for native? Commands are read in-order and RPCs are executed as the stream is read iirc - so properties will be applied after RPCs (not 100% on this)

Reply:
Ah interesting, I think the case of native is that this does happen in a defined way - DataChannel.cpp reads and actions items in order (RPCs are written immediately and properties at end-of-frame) so it seems RPCs should always happen first (chekced with some breakpoints also). Although the behavior is slightly more defined in that RPCs then properties are executed. Unmapped RPC args are sometimes deferred until after property updates though which resolves some of that ambiguity (FObjectReplicator::ReceivedBunch bCanDelayRPCs)

Reply:
Reading more on this, seems like by default, everything is sent immediately, except unreliable net multicast RPCs.
So it looks like native UE is not concerned about replication frame consistency.